packages/vite/src/node/server/hmr.ts

```diff
@@ -486,7 +487,10 @@ async function restartServer(server: ViteDevServer) {
     }
   }
   if (!server.config.server.middlewareMode) {
-    await server.listen(undefined, true)
+    await server.listen(
+      isObject(prevAddress) ? prevAddress.port : undefined,
+      true
+    )
```

Comment:
Together with antfu's suggestion:

```suggestion
    await server.listen(port, true)
```